### ISSUE TITLE
Questions about GPU productivity

### ISSUE THREAD

**Issue by mengqingyu123** [nachiket92/PGP#5]
Hi, thank you so much for your great work and open source. I am very interested in your research, including the previous papers I have read.

When I tried to run the source code of PGP, I found that the working efficiency of the GPU was very low, basically not more than 10%, which caused the training to be extremely slow.

I'm using an NVIDIA RTX2080 GPU, and my computing power is much lower than the computing platform used in your paper. I'm not sure if the low usage is due to GPU memory and I took the liberty to ask this question, thank you very much!

![Screenshot from 2022-01-17 14-44-41](https://user-images.githubusercontent.com/37166419/149721039-e8a1c3c0-0b07-4f91-8974-62bc38d1f244.png)

**Comment by Sungmin-Woo** [comment]
@mengqingyu123 Hi, I have the same issue. Did you find the reason?

**Comment by mengqingyu123** [comment]
@Sungmin-Woo A very simple question, following the steps in the readme, I installed the cpu version of pytorch. After checking I installed the gpu version of pytorch via pip and the problem was solved.

**Comment by Sungmin-Woo** [comment]
@mengqingyu123 Do you mean tensorflow-gpu?
I have installed tensorflow-gpu but the GPU-util is still around 30%~40%.

**Comment by nachiket92** [comment]
I have tested this implementation using a GTX 1080 Ti. Training took about 7 hrs. 
The model is pretty lightweight so the memory usage should not be too high. 
 
Do make sure you install the version of Pytorch that matches your version of  CUDA. 

This is the step that might need to be changed:
conda install pytorch==1.7.1 torchvision==0.8.2 torchaudio==0.7.2 cudatoolkit=10.1 -c pytorch

**Comment by Sungmin-Woo** [comment]
Hi @nachiket92. Yes, my CUDA version is 10.1 as yours.

It works well but my concern is that the Volatile GPU-Util in nvidia-smi is 30%-40%, which means that GPU is used quite inefficiently. Is it not the problem of environment setting? It will be helpful if you check your GPU-Util. Thanks.

**Comment by gilgamsh** [comment]
> Hi @nachiket92. Yes, my CUDA version is 10.1 as yours.
> 
> It works well but my concern is that the Volatile GPU-Util in nvidia-smi is 30%-40%, which means that GPU is used quite inefficiently. Is it not the problem of environment setting? It will be helpful if you check your GPU-Util. Thanks.

My GPU-Util is around about 20% .  Have you sloved the problem?

**Comment by Sungmin-Woo** [comment]
@gilgamsh No, I have not solved yet. As my training time is longer than the author's, it seems that GPU-Util should be higher.

**Comment by gilgamsh** [comment]
> 
@nachiket92  Could you tell us about how much time training a epoch cost? It takes about 15 minutes in my environment( GTX 2080 Ti). Is that a normal number ?

**Comment by nachiket92** [comment]
Sorry for the late response, I was running some tests and working on a fix. 

Here's what seems to be going on:

The bottleneck during training and inference is the clustering step (K-means) to yield the final set of K trajectories from the sampled trajectories. This is done using sklearn.cluster and uses the _CPU_. I do use parallelism across each minibatch using ray.remote, but this is still on CPU.  The large variation in training times reported above are likely due to different numbers of CPU cores on systems. The true fix for this issue would be an implementation of K-means that uses the GPU and has mini-batch support. The closest thing I could find was this really cool implementation: [https://github.com/DeMoriarty/TorchPQ/blob/main/torchpq/clustering/MultiKMeans.py](https://github.com/DeMoriarty/TorchPQ/blob/main/torchpq/clustering/MultiKMeans.py)
However, I couldn't get the same accuracy as using sklearn.cluster. Perhaps I need to run more tests. 

That said, there was a lot of room for improvement in the current code. For example,

This awful nested loop
https://github.com/nachiket92/PGP/blob/546455f2260c22e7f46cf73da8b16f9c04183592/models/decoders/utils.py#L143-L146

Lots of redundant computation in this function
https://github.com/nachiket92/PGP/blob/546455f2260c22e7f46cf73da8b16f9c04183592/models/aggregators/pgp.py#L91

I have restructured and optimized a lot of the decoder code and will push the commits shortly. This should speed up training and inference significantly.

**Comment by gilgamsh** [comment]
> Sorry for the late response, I was running some tests and working on a fix.
> 
> Here's what seems to be going on:
> 
> The bottleneck during training and inference is the clustering step (K-means) to yield the final set of K trajectories from the sampled trajectories. This is done using sklearn.cluster and uses the _CPU_. I do use parallelism across each minibatch using ray.remote, but this is still on CPU. The large variation in training times reported above are likely due to different numbers of CPU cores on systems. The true fix for this issue would be an implementation of K-means that uses the GPU and has mini-batch support. The closest thing I could find was this really cool implementation: https://github.com/DeMoriarty/TorchPQ/blob/main/torchpq/clustering/MultiKMeans.py However, I couldn't get the same accuracy as using sklearn.cluster. Perhaps I need to run more tests.
> 
> That said, there was a lot of room for improvement in the current code. For example,
> 
> This awful nested loop
> 
> https://github.com/nachiket92/PGP/blob/546455f2260c22e7f46cf73da8b16f9c04183592/models/decoders/utils.py#L143-L146
> 
> Lots of redundant computation in this function
> 
> https://github.com/nachiket92/PGP/blob/546455f2260c22e7f46cf73da8b16f9c04183592/models/aggregators/pgp.py#L91
> 
> I have restructured and optimized a lot of the decoder code and will push the commits shortly. This should speed up training and inference significantly.

Thank you very much.

**Comment by nachiket92** [comment]
I've pushed the changes. Training and inference should be faster now.
